### ISSUE TITLE
Handle edgecase with Kotlin boolean types

### ISSUE THREAD

**Issue by amir20001** [jOOQ/jOOQ#8004]
### Steps to reproduce the problem (include example code if possible):

`Foo(val id: String, var isDeleted: Boolean)`

Will use ImmutablePOJOMapperWithParameterNames and try to map to a field called `isDeleted` but the field should be `deleted`

### Versions:

- jOOQ: 3.11
- Java: 8
- Database (include vendor): MySql
- OS: 10.13.6
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thank you very much for your request.

I'm curious: Based on what specific convention do you think that the `is` prefix should somehow be recognised by jOOQ as not being part of the attribute name? jOOQ can recognise it on getters, as that has been common practice for many years according to the JavaBeans convention, but you're not referring to a getter, you're referring to an attribute...

**Comment by amir20001** [comment]
Thanks for getting back to me.
Just to make sure we are on the same page.
 `Foo(val id: String, var isDeleted: Boolean)`

will generate a java class equivalent to 

`
public class Foo {

    private String id;
    private Boolean deleted;

    public Foo(String id, Boolean isDeleted) {
        this.id = id;
        this.deleted = deleted;
    }

    public String getId() {
        return id;
    }

    public Foo setId(String id) {
        this.id = id;
        return this;
    }

    public Boolean isDeleted() {
        return deleted;
    }

    public Foo setDeleted(Boolean deleted) {
        this.deleted = deleted;
        return this;
    }
}
`

The problem that we are running into if im not mistaken, we are reading the property names from the constructor, which is what causes the mismatch.

**Comment by amir20001** [comment]
Unfortunately i can not find anything in the reference guide. All I can find are some stackoverflow posts 
https://stackoverflow.com/questions/43317026/kotlin-naming-convention-for-boolean-returning-methods

**Comment by vojtapol** [comment]
Hi Luke,

This is a widely used feature of Kotlin. 

The corresponding documentation is here:
https://kotlinlang.org/docs/reference/java-interop.html#getters-and-setters

As you know, Kotlin generates getters and setters when you define a class using this syntax:

```kotlin
data class User(val name: String, val age: Int)
```

This generates `getName()` and `getAge()` methods in the bytecode.

The problem is though that often but not always, boolean getters in Java start with `is` instead of `get`. JetBrains had to come up with a way to let users choose whether they want the generated getter to start with `is` or with `get` while maintaining the concise syntax.

The solution they chose was that if a Kotlin *property name* (which is also the primary constructor parameter name) starts with `is` and it is of type `Boolean` the generated getter will not prepend `get` and just use the property name. In all other cases, `get` is prepended.

---

To resolve this ticket, I think we should adjust the constructor parameter name matcher to accept both `is<name>` and `<name>` if it is a boolean. I don't think this change would break existing Java code as the only case where it would break is if someone defined a constructor like this:
```java
public User(String name, boolean isActive, boolean active)
```

If this is a concern, we could somehow enable this fix only when mapping a Kotlin class.

**Comment by lukaseder** [comment]
> The problem is though that often but not always, boolean getters in Java start with is instead of get. JetBrains had to come up with a way to let users choose whether they want the generated getter to start with is or with get while maintaining the concise syntax.

That ... makes sense but sucks :( They should have solved this with an annotation I guess. Given that they support named parameters, how could they abuse the property name for this non-essential feature? I'm really curious about the funky edge cases they ran into, and which they're solving with this particular approach.

> I don't think this change would break existing Java code

No, it doesn't. Your own implementation savely guards this particular logic from Java code (from https://github.com/jOOQ/jOOQ/issues/7324):

```java
        // [#7324] Map immutable Kotlin classes by parameter names if kotlin-reflect is on the classpath
        if (Tools.isKotlinAvailable()) {
            try {
                Reflect jvmClassMappingKt = Tools.ktJvmClassMapping();
                Reflect kClasses = Tools.ktKClasses();

                Object klass = jvmClassMappingKt.call("getKotlinClass", type).get();
                Reflect primaryConstructor = kClasses.call("getPrimaryConstructor", klass);

                // It is a Kotlin class
                if (primaryConstructor.get() != null) {
```

All adaptations will be made within that scope.

> If this is a concern, we could somehow relax the name matching only when mapping a Kotlin class.

Yes, of course! This lunacy must not affect Java developers :-) (We already have to deal with the useless `is` / `get` duality ourselves)

Anyway, enough of the ranting. Obviously, if that's how things work there, we must support it. Anyone interested in providing a PR?

**Comment by vojtapol** [comment]
I am not a big fan of the solution they chose myself. It is what it is though so we should handle it.

I would be willing to work on the PR but can't commit to it at the moment. If anyone is willing to jump in and do it, consider it still unclaimed for now.

**Comment by lukaseder** [comment]
The current fix only works with primitive type `boolean` (Java) / `Boolean` (Kotlin), not with the wrapper type `java.lang.Boolean` (Java) / `Boolean?` (Kotlin).

Will fix this as well

**Comment by lukaseder** [comment]
@amir20001 @vojtapol while we're at it, I'm thinking of supporting a few out of the box Kotlin extension functions to provide a more idiomatic jOOQ experience to Kotlin users, see #6256.

Do you have such utilities on your side? Anything you'd be interested in sharing?

Also, what would be the most idiomatic approach here? Provide these utilities as top-level functions directly in packages? Or in companion objects? Or something else?

**Comment by lukaseder** [comment]
Another, related fix when data classes have generic type variables:
https://github.com/jOOQ/jOOQ/commit/f2ad98e7f7314b064a982e8748e1a0330b902141

This fixes fetching into a class like this one:

```kotlin
    data class GenericDataClass<T, U : Long>(val group: T, val count: U)

    @Test
    fun testDataClassMapperWithGenerics() {
        val result =
        ctx().select(DSL.one().`as`("group"), DSL.two().`as`("count"))
                .fetchOneInto(GenericDataClass::class)

        assertEquals(1, result.group);
        assertEquals(2, result.count);
    }
```

**Comment by lukaseder** [comment]
Huh, our CI was misconfigured, so the animal sniffer plugin didn't detect the flaw. We cannot use `Class.getTypeName()` in our code, as we're still supporting Java 6-7. This line breaks our release build for Java 6:

https://github.com/jOOQ/jOOQ/blob/version-3.12.0-RC1/jOOQ/src/main/java/org/jooq/impl/DefaultRecordMapper.java#L419-L420

Luckily, we don't need to call `getTypeName()`. `getName()` is sufficient for the below checks.

**Comment by lukaseder** [comment]
Here's another fun case:

```kotlin
data class Test(
    var x: Boolean,
    var isX: Boolean
)

fun main() {
    println(Test(true, false))
}
```

It produces:

```
Exception in thread "main" java.lang.ClassFormatError: Duplicate method name "setX" with signature "(Z)V" in class file Test
 at java.lang.ClassLoader.defineClass1 (:-2) 
 at java.lang.ClassLoader.defineClass (:-1) 
 at java.security.SecureClassLoader.defineClass (:-1) 
```

I'm thinking that this clever "feature" of kotlin is really a mistake that should be fixed in kotlin:
https://youtrack.jetbrains.com/issue/KT-63097

And the best fix here might be to revert #8004. Though, I still haven't decided.
